File: app/src/main/java/cn/ifafu/ifafu/mvp/score_filter/ScoreFilterModel.java
package cn.ifafu.ifafu.mvp.score_filter;

import android.content.Context;

import java.util.List;

import cn.ifafu.ifafu.dao.ScoreDao;
import cn.ifafu.ifafu.data.entity.Score;
import cn.ifafu.ifafu.data.local.DaoManager;
import cn.ifafu.ifafu.base.BaseModel;
import io.reactivex.Observable;

public class ScoreFilterModel extends BaseModel implements ScoreFilterConstant.Model {

    private ScoreDao scoreDao = DaoManager.getInstance().getDaoSession().getScoreDao();

    ScoreFilterModel(Context context) {
        super(context);
    }

    @Override
    public Observable<List<Score>> getScoresFromDB(String year, String term) {
        return Observable.fromCallable(() -> {
            if (year.equals("全部") && term.equals("全部")) {
                return repository.getAllScores();
            } else if (year.equals("全部")) {
                return repository.getScoresByTerm(term);
            } else if (term.equals("全部")) {
                return repository.getScoresByYear(year);
            } else {
                return repository.getScores(year, term);
            }
        });
    }

    @Override
    public void save(Score score) {
        scoreDao.insertOrReplace(score);
    }

}

File: app/src/main/java/cn/ifafu/ifafu/mvp/score_filter/ScoreFilterModel.kt
package cn.ifafu.ifafu.mvp.score_filter

import android.content.Context
import cn.ifafu.ifafu.base.BaseModel
import cn.ifafu.ifafu.data.RepositoryImpl
import cn.ifafu.ifafu.entity.Score
import io.reactivex.Observable
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class ScoreFilterModel constructor(context: Context) : BaseModel(context), ScoreFilterConstant.Model {
    override fun getScoresFromDB(year: String, term: String): Observable<List<Score>> {
        return Observable.fromCallable {
            if (year == "全部" && term == "全部") {
                return@fromCallable repository.getAllScores()
            } else if (year == "全部") {
                return@fromCallable repository.getScoresByTerm(term)
            } else if (term == "全部") {
                return@fromCallable repository.getScoresByYear(year)
            } else {
                return@fromCallable repository.getScores(year, term)
            }
        }
    }

    override fun save(score: Score) {
        GlobalScope.launch(Dispatchers.IO) {
            RepositoryImpl.saveScore(score)
        }
    }
}